src/plugins/plugin-http.js

```diff
@@ -0,0 +1,137 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+var shimmer = require('shimmer');
+var url = require('url');
+var isString = require('is').string;
+var merge = require('lodash.merge');
+var httpAgent = require('_http_agent');
+var constants = require('../constants.js');
+var TRACE_AGENT_REQUEST_HEADER = require('../constants.js').TRACE_AGENT_REQUEST_HEADER;
+
+function isTraceAgentRequest (options) {
+  return options && options.headers &&
+    !!options.headers[TRACE_AGENT_REQUEST_HEADER];
+}
+
+function getSpanName(requestOptions) {
+  if (isString(requestOptions)) {
+    requestOptions = url.parse(requestOptions);
+  }
+  // c.f. _http_client.js ClientRequest constructor
+  return requestOptions.hostname || requestOptions.host || 'localhost';
+}
+
+function setTraceHeader(parsedOptions, context, api) {
+  if (context) {
+    return merge(parsedOptions, {
+      headers: {
+        [constants.TRACE_CONTEXT_HEADER_NAME]: context
+      }
+    });
+  }
+  return parsedOptions;
+}
+
+function extractUrl(parsedOptions) {
+  var uri = parsedOptions;
+  var agent = parsedOptions._defaultAgent || httpAgent.globalAgent;
+  return isString(uri) ? uri :
+    (parsedOptions.protocol || agent.protocol) + '//' +
+    (parsedOptions.hostname || parsedOptions.host || 'localhost') +
+    ((isString(parsedOptions.port) ? (':' + parsedOptions.port) : '')) +
+    (parsedOptions.path || parsedOptions.pathName || '/');
+}
+
+function parseRequestOptions(requestOptions) {
+  return isString(requestOptions) ?
+    merge({headers: {}}, url.parse(requestOptions)) :
+    merge({headers: {}}, requestOptions);
+}
+
+function patchedHTTPRequest(requestOptions, callback, request, api) {
+  if (!requestOptions) {
+    return request.call(request, requestOptions, callback);
+  } else if (isTraceAgentRequest(requestOptions)) {
+    return request.call(request, requestOptions, callback);
+  } else if (!api.getRootSpan()) {
+    if (isString(requestOptions)) {
+      requestOptions = url.parse(requestOptions);
+    }
+    // What's the new logger target?
+    // console.log('Untraced http uri:', extractUrl(requestOptions));
+    return request.call(request, requestOptions, callback);
+  }
+  var parsedOptions = parseRequestOptions(requestOptions);
+  var uri = extractUrl(parsedOptions);
+  var requestLifecycleSpan = api.createChildSpan({name: getSpanName(parsedOptions)});
+  if (!requestLifecycleSpan) {
```

Comment:
`requestLifecycleSpan` is guaranteed to be non-null because `api.createChildSpan` depends on `api.getRootSpan`, the latter already being verified as non-null earlier. Just FYI... I'm not 100% if it is a best practice to check anyway.